### ISSUE TITLE
Don't load static assets via PHP (for performance and Cloud compatibility)

### ISSUE THREAD

**Issue by timkelty** [solspace/craft3-freeform#1278]
Freeform [loads front-end resources via PHP](https://github.com/solspace/craft-freeform/blob/v5/packages/plugin/src/controllers/ResourcesController.php) (`Response::sendFile`), which is very inefficient. This is exacerbated on infrastructure Craft Cloud, where PHP can never send files, and will result in a redirect to and s3 upload.

Static front-end resources should always be defined via Asset Bundles, allowing them be be served directly.

There are some guidelines [here](https://craftcms.com/knowledge-base/cloud-plugin-development#other-best-practices) regarding this, but we will continue to improve them.

**Comment by gustavs-gutmanis** [comment]
Hi @timkelty 

Can I ask you a couple of questions just so I understand completely?

### Current Implementation
Currently, Freeform allows you to pick one of three options

1. serve Freeform JS and CSS files from a static URLs 
    * `your.site/freeform/plugin.js`
    * `your.site/freeform/plugin.css`
  
      This approach uses `$response->sendFile()` as you mentioned, which sends `Cache-Control`, `ETag` and `Last-Modified` headers, so that the browser never reloads these scripts unless they have changed. The files don't actually exist, but the end-user does not see the difference.

2. serve Freeform JS and CSS files using `\Craft::$app->assetManager->getPublishedUrl()`. This approach does what your PR #1286 is trying to do.

3. serve Freeform JS and CSS files inline in the form via `<script>` tags.

### Question
If we're assuming that the best approach is the **2nd** one, would it be sufficient for people to use that via setting it in their settings? (We could default to it from now on)

What harm does the **1st** approach do, given that it is cached by the browser? The file is read by PHP which should be possible even in Cloud, and then it is being sent as a response with the correct headers?

Assuming that this isn't possible with Craft Cloud, people should still be able to use the **2nd** approach and serve the files via published assets, right?

If that's the case, then we could see if Craft Cloud is being used and remove the option from the settings entirely. Would that work?

**Comment by timkelty** [comment]
Certainly!

> If we're assuming that the best approach is the 2nd one, would it be sufficient for people to use that via setting it in their settings?

Yep! I didn't realize there were options for how these were rendered.

> What harm does the 1st approach do, given that it is cached by the browser? The file is read by PHP which should be possible even in Cloud, and then it is being sent as a response with the correct headers?

In Craft Cloud (or other similar Lambda-based infrastructures), PHP is dedicated to rendering "traditional" web requests, and can't really send files. That said, we do have a workaround in place for when this is necessary (e.g. a controller needs to do some "work" and send a file, like generating a PDF). The workaround uploads to a presigned s3 URL and then serves a redirect to it, so, is not very efficient for something like this. Furthermore, we have to override any `Cache-control` headers to make it not cacheable, because the redirected URL is short-lived, so it wont end up being cached by the browser.

That aside, in this case, theres just no _reason_ to rely on the inherently volatile browser cache, when all you're doing is serving a static file. Imagine a flood of millions of requests – even with browser cache, each one would have to go through PHP and serve that file, when it just as easily bypassed PHP entirely and hit a CDN directly.

> If that's the case, then we could see if Craft Cloud is being used and remove the option from the settings entirely. Would that work?

Yep, that should work. You can test for `craft\cloud\Helper::isCraftCloud`, which will return `true` if currently running in Craft Cloud.

---

Reading the settings description, I see now that you're defaulting/recommending option 1 specifically because of the cache headers you're adding. It is worth noting that Craft _will_ send cache headers for initial cpresources request, if the app is using the `GeneralConfig::buildId` config setting: https://github.com/craftcms/cms/blob/5.x/src/web/Application.php#L550-L554

Even without that, generally it would be the responsibility of the web server (nginx, etc) to set cache headers for subsequent requests to the static resource.

In general, I would consider removing/deprecating option 1 (not just for Cloud) in future version. It seems like more trouble that it is worth.

**Comment by gustavs-gutmanis** [comment]
Thank you so much for the in-depth reply @timkelty.
It seems like the **1st** option is entirely useless. We will remove it and switch to using bundled `cpresources` instead by default.

**Comment by timkelty** [comment]
Happy to test whenever you have something out!

**Comment by kjmartens** [comment]
@timkelty Does using the "As Files" option for the _Freeform Script Insert Type_ setting currently work as is? 🙂 

<img width="336" alt="Screenshot 2024-05-22 at 10 46 04 AM" src="https://github.com/solspace/craft-freeform/assets/1821674/512632f2-8b9f-4bb9-8fec-b4f0937d5bee">

**Comment by timkelty** [comment]
@kjmartens it doesn't!

```
Invalid Argument – [yii\base\InvalidArgumentException](https://www.yiiframework.com/doc-2.0/yii-base-invalidargumentexception.html)
The file or directory to be published does not exist: https://craftcom.ddev.site/cpresources/f57e6f09/js/scripts/front-end/plugin/freeform.js?v=1715958477
```

It looks like [here](https://github.com/solspace/craft-freeform/blob/v5/packages/plugin/src/Events/Forms/RenderTagEvent.php#L172), `$filePath` is a full URL already, when it should be a path (and confusingly, `$url` is `plugin/freeform.js`)

In this case, the `getPublishedUrl` call needs to look like I have it in this PR: https://github.com/solspace/craft-freeform/pull/1286/files#diff-497338c84b7afba080358b8c528398f1e721a2abcf40efc8dcb3bdb85c85ed53R377-R381

**Comment by gustavs-gutmanis** [comment]
Hey @timkelty, could you please take a look at #1299?

**Comment by timkelty** [comment]
My tests look great, @gustavs-gutmanis.

**Comment by kjmartens** [comment]
This is now resolved in Freeform 5.3.0+ 🙂 

Thanks for your help on this one @timkelty 🤗

**Comment by timkelty** [comment]
Thank you!
